File: split/showcontext.pas
procedure showcontext;
label
  30;
var
  oldsetting: 0..21;
  nn: integer;
  bottomline: boolean;
  {315:}
i: 0..bufsize;
  j: 0..bufsize;
  l: 0..halferrorline;
  m: integer;
  n: 0..errorline;
  p: integer;
  q: integer;
  {:315}
begin
  baseptr := inputptr;
  inputstack[baseptr] := curinput;
  nn := -1;
  bottomline := false;
  while true do
  begin
    curinput := inputstack[baseptr];
    if (curinput.statefield <> 0) then
      if (curinput.namefield > 17) or (baseptr = 0) then
        bottomline := true;
    if (baseptr = inputptr) or bottomline or (nn < eqtb[5317].int) then{312:}

    begin
      if (baseptr = inputptr) or (curinput.statefield <> 0) or (curinput.indexfield <> 3) or (curinput.locfield <> 0) then
      begin
        tally := 0;
        oldsetting := selector;
        if curinput.statefield <> 0 then
        begin
          {313:}
if curinput.namefield <= 17 then
            if (curinput.namefield = 0) then
              if baseptr = 0 then
                printnl(574)
              else
                printnl(
                  575)
            else
            begin
              printnl(576);
              if curinput.namefield = 17 then
                printchar(42)
              else
                printint(curinput.namefield - 1);
              printchar(62);
            end
          else
          begin
            printnl(577);
            printint(line);
          end;
          printchar(32){:313}
;{318:}

          begin
            l := tally;
            tally := 0;
            selector := 20;
            trickcount := 1000000;
          end;
          if buffer[curinput.limitfield] = eqtb[5311].int then
            j := curinput.limitfield
          else
            j := curinput.limitfield + 1;
          if j > 0 then
            for i := curinput.startfield to j - 1 do
            begin
              if i = curinput.locfield then
              begin
                firstcount := tally;
                trickcount := tally + 1 + errorline - halferrorline;
                if trickcount < errorline then
                  trickcount := errorline;
              end;
              print(buffer[i]);
            end{:318}
;
        end
        else
        begin
          {314:}
case curinput.indexfield of
            0: printnl(578);
            1, 2: printnl(579);
            3: if curinput.locfield = 0 then
                printnl(580)
              else
                printnl(581);
            4: printnl(582);
            5:
            begin
              println;
              printcs(curinput.namefield);
            end;
            6: printnl(583);
            7: printnl(584);
            8: printnl(585);
            9: printnl(586);
            10: printnl(587);
            11: printnl(588);
            12: printnl(589);
            13: printnl(590);
            14: printnl(591);
            15: printnl(592);
            others: printnl(63)
          end{:314}
;
          {319:}
begin
            l := tally;
            tally := 0;
            selector := 20;
            trickcount := 1000000;
          end;
          if curinput.indexfield < 5 then
            showtokenlist(curinput.startfield, curinput.locfield, 100000)
          else
            showtokenlist(mem[curinput.startfield].hh.rh,
              curinput.locfield, 100000){:319}
;
        end;
        selector := oldsetting;
        {317:}
if trickcount = 1000000 then
        begin
          firstcount := tally;
          trickcount := tally + 1 + errorline - halferrorline;
          if trickcount < errorline then
            trickcount := errorline;
        end;
        if tally < trickcount then
          m := tally - firstcount
        else
          m := trickcount - firstcount;
        if l + firstcount <= halferrorline then
        begin
          p := 0;
          n := l + firstcount;
        end
        else
        begin
          print(277);
          p := l + firstcount - halferrorline + 3;
          n := halferrorline;
        end;
        for q := p to firstcount - 1 do
          printchar(trickbuf[q mod errorline]);
        println;
        for q := 1 to n do
          printchar(32);
        if m + n <= errorline then
          p := firstcount + m
        else
          p := firstcount + (errorline - n - 3);
        for q := firstcount to p - 1 do
          printchar(trickbuf[q mod errorline]);
        if m + n > errorline then
          print(277){:317}
;
        nn := nn + 1;
      end;
    end{:312}

    else if nn = eqtb[5317].int then
    begin
      printnl(277);
      nn := nn + 1;
    end;
    if bottomline then
      goto 30;
    baseptr := baseptr - 1;
  end;
  30:
    curinput := inputstack[inputptr];
end;

File: cpp/showcontext.cpp
#include "showcontext.h"
#include "printnl.h"
#include "printint.h"
#include "printchar.h"
#include "print.h"
#include "println.h"
#include "printcs.h"
#include "showtokenlist.h"
#include "texte.h"

static void begin_pseudoprint(int l)
{
	l = tally;
	tally = 0;
	selector = pseudo;
	trickcount = 1000000;
}

static void set_trick_count(void)
{
	firstcount = tally; 
	trickcount = tally+1+errorline-halferrorline;
	if (trickcount < errorline)
		trickcount = errorline;
}

void showcontext(void)
{
	int l;
	baseptr = inputptr;
	inputstack[baseptr] = curinput; 
	int nn = -1;
	bool bottomline = false;
	while (true) 
	{
		curinput = inputstack[baseptr];
		if (state)
			if (txt(name) > 17 || baseptr == 0)
				bottomline = true;
		if (baseptr == inputptr || bottomline || nn < error_context_lines())
		{
			if (baseptr == inputptr || state != token_list || token_type != backed_up || loc)
			{
				tally = 0;
				oldsetting = selector;
				if (state)
				{
					if (txt(name) <= 17)
						if (terminal_input(name))
							printnl(baseptr == 0 ? "<*>" : "<insert> ");
						else
						{
							printnl("<read ");
							if (txt(name) == 17)
								printchar('*');
							else
								printint(txt(name)-1);
							printchar('>');
						}
					else
					{
						printnl("l.");
						printint(line);
					}
					printchar(' ');
					begin_pseudoprint(l);
					int j;
					if (buffer[limit] == end_line_char())
						j = limit;
					else
						j = limit+1;
					if (j > 0)
						for (int i = start; i < j; i++)
						{
							if (i == loc)
								set_trick_count();
							printchar(buffer[i]);
						}
				}
				else
				{
					switch (token_type)
					{
						case parameter: 
							printnl("<argument> ");
							break;
						case u_template:
						case v_template:
							printnl("<template> ");
							break;
						case backed_up: 
							if (loc == 0)
								printnl("<recently read> ");
							else
								printnl("<to be read again> ");
							break;
						case inserted: 
							printnl("<inserted text> ");
							break;
						case macro:
							println();
							printcs(txt(name));
							break;
						case output_text: 
							printnl("<output> ");
							break;
						case every_par_text: 
							printnl("<everypar> ");
							break;
						case every_math_text: 
							printnl("<everymath> ");
							break;
						case every_display_text: 
							printnl("<everydisplay> ");
							break;
						case every_hbox_text: 
							printnl("<everyhbox> ");
							break;
						case every_vbox_text: 
							printnl("<everyvbox> ");
							break;
						case every_job_text: 
							printnl("<everyjob> ");
							break;
						case every_cr_text: 
							printnl("<everycr> ");
							break;
						case mark_text: 
							printnl("<mark> ");
							break;
						case write_text: 
							printnl("<write> ");
							break;
						default: 
							printnl("?");
					}
					begin_pseudoprint(l);
					if (token_type < macro)
						showtokenlist(start, loc, 100000);
					else
						showtokenlist(link(start), loc, 100000);
				}
				selector = oldsetting;
				if (trickcount == 1000000)
					set_trick_count();
				int m;
				if (tally < trickcount)
					m = tally - firstcount;
				else
					m = trickcount-firstcount;
				int p, n;
				if (l+firstcount <= halferrorline)
				{
					p = 0;
					n = l+firstcount;
				}
				else
				{
					print("...");
					p = l+firstcount-halferrorline+3;
					n = halferrorline;
				}
				for (int q = p; q < firstcount; q++)
					printchar(trickbuf[q%errorline]);
				println();
				for (int q = 0; q < n; q++)
					printchar(' ');
				if (m + n <= errorline)
					p = firstcount+m;
				else
					p = firstcount+errorline-n-3;
				for (int q = firstcount; q < p; q++)
					printchar(trickbuf[q%errorline]);
				if (m+n > errorline)
					print("...");
				nn++;
			}
		}
		else 
			if (nn == error_context_lines())
			{
				printnl("...");
				nn++;
			}
		if (bottomline)
			break;
		baseptr--;
	}
	curinput = inputstack[inputptr];
}
